libs/collectives/src/detail/barrier_node.cpp

```diff
@@ -196,13 +210,20 @@ namespace hpx { namespace lcos { namespace detail {
         }
 
         hpx::intrusive_ptr<barrier_node> this_(this);
+
         // Once we know that all our children entered the barrier, we flag ourself
-        return hpx::when_all(futures).then(
-            hpx::launch::sync, [this_ = std::move(this_)](hpx::future<void> f) {
+        auto result = hpx::when_all(futures).then(
+            hpx::launch::sync, [this_](hpx::future<void> f) {
                 // Trigger possible errors...
                 f.get();
                 return this_->gather_promise_.get_future();
             });
+
+        // keep everything alive until future has become ready
+        traits::detail::get_shared_state(result)->set_on_completed(
+            [this_ = std::move(this_)] {});
```

Comment:
How much longer does this keep `this` alive than just capturing it above in the `then` continuation? Is the point that it has to live past the future actually being set (i.e. to the calling scope of the continuation), and why does that make a difference?

Reply:
Fair point. I'll go back and check again. I might have been a bit overzealous here while I was doing the investigation.

Reply:
My idea was to keep it alive as long as the returned future lives. The captured `this_` in the `.then()` expression lives as long as the future the continuation was attached to only.

Reply:
I did look again. I think my changes are valid. We could remove the capture of `this_` for the continuations attached using `.then()`, however.